Gutenboarding: improve preview quality

simison opened the pull request:
Resolves https://github.com/Automattic/wp-calypso/issues/43103

#### Changes proposed in this Pull Request

Hi,

This PR does two things:

- extends the functionality of the preview generator scripts so that we only have to install the dependencies and run the script (rather than manually create json, edit package.json and install required deps)
- generates higher resolution thumbails

**Before (600w)**
![doyle_doyle_alves-600](https://user-images.githubusercontent.com/6458278/84220057-3f20d280-ab15-11ea-9253-1bcf404924e2.jpg)


**After (700w)**
![doyle_doyle_alves-700](https://user-images.githubusercontent.com/6458278/84220064-42b45980-ab15-11ea-8fcb-fee97a027241.jpg)

#### Testing instructions

Follow the instructions in `bin/generate-gutenboarding-design-thumbnails.js`

Make sure the thumbnails are generated and have the dimensions

Take a peek at http://calypso.localhost:3000/new/design 

Does it look okay?

andrewserong reviewed (commented):
I tested this out to see if it resolves #43103, and at most smaller viewports (up until around 1400px wide) I can't see much of a quality difference. However at full screen widths with a high res monitor (>2000px) I can see the blurriness issue in #43103, and this change makes it a little bit better, but the pixel width of the images still seems to be the issue, as opposed to the JPEG quality.

Since at smaller resolutions / viewports the quality seems to be much the same between master and this PR, what if we upped the outputWidth to something more like 900 or 1000px, and reduced the JPEG quality down to, say, 70-80%? With the quality up at 98% in this change, it increases the file-size quite dramatically (~1.4mb take all together), so hopefully we can find a good trade-off between file-size and image quality. The design of the step is a bit of a challenge for these screenshots, because the tiles are very much thumbnails at small viewports, but at the wider viewports, these images end up being nearly a full page preview in size. Longer-term I guess we could produce different images to target different viewports?

Other than that, I saw a few visual differences in the screenshots (not quite sure what's causing it!). In this PR:

* The background image in Cassel isn't displayed
* The Latest Posts in Edison aren't rendered 
* The CTA button text in Overton appears to be grey instead of white, but that's also the case in the preview, so I'm not quite sure what's going on there

andrewserong commented on bin/generate-gutenboarding-design-thumbnails.js:
```diff
@@ -37,9 +38,44 @@ const getDesignUrl = ( design ) => {
+		.map( ( line ) => line.trim() )
+		.filter( Boolean )
+		.join( '' );
+	featuredDesignsObject = eval( '(' + trimmedString + ')' ).featured;
```
Clever regex use here! To avoid having to `eval` the string, would it be worth it to move the list of availableDesigns out of the `ts` file and into a `json` file that we can import from `available-designs.ts`, and use `JSON.parse` here? Alternatively, I just read about this approach to avoiding `eval` and [using window.Function()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#Never_use_eval!) instead.

simison replied:
Good point about going back to json file! :-) I was hoping we wouldn't need to run this script more than a few times and that we'd work on mshots instead of improving this script. 🙃 

https://github.com/Automattic/wp-calypso/pull/40632#discussion_r402218015

ramonjd replied:
> to avoid having to eval the string, would it be worth it to move the list of availableDesigns out of the ts file and into a json file that we can import from available-designs.ts

Totally worth it. And it will make this rush job looks less, well, like a rush job. 👍

ramonjd committed:
These scripts simplify the steps required to install the dependencies to create the thumbnails:

- a custom yarn install script that installs chromium and screenshot libraries
- extracting the designs to a json file so that we can use the raw data everywhere

ramonjd committed:
Regenerating at 900w at 72% quality

ramonjd commented:
I've rebased, created a json for the available designs config and tweaked the image variables according to @andrewserong's suggestions

>  I saw a few visual differences in the screenshots

I think many of those differences are not there in this round of exports 🤷 I get the impression that loading the demos sites might be a bit flakey or it might even have been my internet connection or the direction of the wind.

andrewserong commented on client/landing/gutenboarding/available-designs.ts:
```diff
@@ -8,147 +8,14 @@ import { addQueryArgs } from '@wordpress/url';
  */
 import { isEnabled } from '../../config';
 import type { Design } from './stores/onboard/types';
+const availableDesignsConfig = require( './available-designs-config.json' );
```
I was looking into whether we could swap this out for:

```
import availableDesignsConfig from './available-designs-config.json';
```

However it then raises a type error in `const availableDesigns: Readonly< AvailableDesigns > = `availableDesignsConfig` because it looks at the heading and base fonts as strings, instead of literals of the type `Font`. It seems that importing JSON in TypeScript doesn't allow for importing as literals, as raised in this issue: https://github.com/microsoft/TypeScript/issues/26552

As a workaround we can leave this import using `require` (which I think means that TypeScript isn't following what's happening with the `availableDesignsConfig variable)  or mess around with our types so that `AvailableDesigns` doesn't expect a FontPair with `headings` and `base` of type `Font` but instead a font pair of strings.

Since this is working as is, and we ultimately want to move this JSON to an endpoint anyway, I think I'd leave it with `require`, unless anyone has strong opinions on it.

ramonjd replied:
Thanks for investigating this @andrewserong !!!

andrewserong reviewed (approved):
Images are looking great to me, higher res, with reduced file-size for most of them 👍 

The design thumbnails dependencies installation script and thumbnail generator are working well for me too.

I left a note about importing the JSON file and dealing with types, which is a bit of complexity I didn't know about before suggesting we move to the JSON file 😅, but I think the change is looking good. We'll ultimately be moving the JSON to an endpoint eventually anyway, so I'm not sure how beneficial it is to be too strict on the types for `availableDesigns` 🤔 

At any rate, I think this is looking good to merge, and we can always clean up the types in a follow up if we need to!

ramonjd commented:
> We'll ultimately be moving the JSON to an endpoint eventually anyway, so I'm not sure how beneficial it is to be too strict on the types for availableDesigns

Agree here. Unless the client really needs to know what to expect from the endpoint. Might be good to catch unexpected responses maybe?

simison commented:
We can type API responses too, and I'd assume we can type the json file outside it without having it as ts-file.

I'm good having this as-is now so that working with these gets easier, since it seems we're needing the script more often than I thought originally. :-)

simison reviewed (commented):
✅ Looking good. 👍(Haven't tested)  I think we can ship as long as we're not introducing typescript errors.
